tool/bindings_generator/js_types_least_common_supertypes.dart

```diff
@@ -0,0 +1,632 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// Updated by /tool/update_bindings.dart. Do not modify by hand.
+
+const Map<String, Map<String, Set<String>>> jsTypesLeastCommonSupertypes = {
```

Comment:
I'm not sure I have the full context, but I am wondering whether we can use a different model for computing the least upperbound that doesn't require generating the expansion of the LUB for each pair of types here.

I believe the JS types form a simple tree-like structure in their subtyping relation, so I am hoping if we use that tree structure instead, it may simplifies the generator and the rules here.

Reply:
There isn't much more context than what's here. :)

A few comments:
- JS types are indeed a tree, but we may or may not change that in the future, but we could just move up the assertions for now and revisit if the type hierarchy gets complex. I'll add that revision.
- We can definitely just output a type hierarchy, and that's what I initially did, but it slows down the generator. You're going to be computing a LUB every time you come across a union rather than looking at a cache. Practically, it might not make much of a difference in terms of performance. I'll add the revision and see what it looks like.

Reply:
Thanks! Regarding performance: We could dynamically cache results too, that could potentially be a middle ground to get something that is more compact but still efficient.

Reply:
Done, PTAL. Some comments on the diff:

- I don't think we need to cache results, the hierarchy is quite flat, the LUB calculation is quick, and there aren't *that* many unions.
- I had to add some additional changes to make this work by moving the updates of partial interfaces to another phase. For context, when we parse the AST and compute an intermediate representation, we resolve partial interfaces by computing unions of their type signatures. However, we haven't visited all the types yet, so this calculation might lead to incorrect results. So, this resolution of partial interfaces is moved to the step after we compute the intermediate representation, but before the translation/emission.
- I added some extra support for other types in the union calculation e.g. interfaces and updated some missing types in our type aliases.

Thanks!